File: python_programmers_school/level 0/분수의 덧셈.py
from fractions import Fraction

def solution(denum1, num1, denum2, num2):
    fraction = Fraction(denum1 * num2 + denum2 * num1) / Fraction(num1 * num2)     # 부동소수점 문제때문에 Fraction으로 한번에 묶지 않고 나누어서 묶어줌.
    return [fraction.numerator, fraction.denominator]

print(solution(9, 2, 1, 3))

# Fraction 모듈 더 간단히 이용하기
# from fractions import Fraction

# def solution(denum1, num1, denum2, num2):
#     answer = Fraction(denum1, num1) + Fraction(denum2, num2)
#     return [answer.numerator, answer.denominator]


# math 모듈 이용하기
# import math

# def solution(denum1, num1, denum2, num2):
#     denum = denum1 * num2 + denum2 * num1
#     num = num1 * num2
#     gcd = math.gcd(denum, num)     math.gcd는 denum과 num 사이의 최대공약수를 찾아준다.
#     return [denum//gcd, num//gcd]     


# 최대공약수 구해서 분모 분자 숫자를 그것으로 나누어서 푸는 방법
# def solution(denum1, num1, denum2, num2):
#     answer = []
#     s = 0

#     denum0 = (denum1*num2) +(denum2*num1)
#     num0 = num1*num2

#     for i in range(min(denum0,num0),0,-1):
#         if denum0%i == 0 and num0%i == 0:
#             s = i
#             break

#     denum0 /= s
#     num0 /= s
#     answer.append(denum0)
#     answer.append(num0)

#     return answer

File: javascript_programmers_school/level 0/분수의 덧셈.js
// 유클리드 호제법 이용해 최대공약수 구하기
// 유클리드 호제법에서 꼭 a가 b보다 큰 수여야 하는 것은 아니다. 크기 상관 없이 적어줄 수 있으므로
// Math.max와 min을 사용할 필요 X
function GCD(a, b) {
  if (a % b === 0) {
    return b;
  } else {
    return GCD(b, a % b);
  }
}

function solution(numer1, denom1, numer2, denom2) {
  var result = [numer1 * denom2 + numer2 * denom1, denom1 * denom2];
  var gcd = GCD(Math.max(result[0], result[1]), Math.min(result[0], result[1]));
  result[0] /= gcd;
  result[1] /= gcd;
  return result;
}

// 최대공약수 구하는 코드를 좀 더 간단하게 나타낸 풀이 (삼항 연산자로 재귀 작성)
function fnGCD(a, b) {
  return a % b ? fnGCD(b, a % b) : b;
}

function solution(denum1, num1, denum2, num2) {
  let denum = denum1 * num2 + denum2 * num1;
  let num = num1 * num2;
  let gcd = fnGCD(denum, num); //최대공약수

  return [denum / gcd, num / gcd];
}
